From: Iain Lane <iainl@gnome.org>
Date: Mon, 11 Feb 2019 12:02:20 +0000
Subject: tests: Install the slow-connect-preload.so library and use it

The gsocketclient-slow test needs this, otherwise connect() succeeds
immeidately and the test fails, because it is checking that cancellation
works. We weren't installing it for installed tests.

(cherry picked from commit a380ddada199d0d88c31dadea3ef09270d47be8d)
---
 gio/tests/meson.build | 13 +++++++++++++
 meson.build           |  3 +++
 template-tap.test.in  |  2 +-
 3 files changed, 17 insertions(+), 1 deletion(-)

diff --git a/gio/tests/meson.build b/gio/tests/meson.build
index 2af1446..c53c445 100644
--- a/gio/tests/meson.build
+++ b/gio/tests/meson.build
@@ -140,11 +140,16 @@ if host_machine.system() != 'windows'
           'slow-connect-preload.c',
           name_prefix : '',
           dependencies: cc.find_library('dl'),
+          install_dir : installed_tests_execdir,
+          install: installed_tests_enabled
         )
       ],
       'env' : {
         'LD_PRELOAD': '@0@/slow-connect-preload.so'.format(meson.current_build_dir())
       },
+      'installed_tests_env' : {
+        'LD_PRELOAD': '@0@/slow-connect-preload.so'.format(installed_tests_execdir)
+      },
       'suite': ['flaky'],
     },
     'gschema-compile' : {'install' : false},
@@ -593,11 +598,19 @@ foreach test_name, extra_args : gio_tests
   source = extra_args.get('source', test_name + '.c')
   extra_sources = extra_args.get('extra_sources', [])
   install = installed_tests_enabled and extra_args.get('install', true)
+  installed_tests_env = extra_args.get('installed_tests_env', {})
 
   if install
     test_conf = configuration_data()
     test_conf.set('installed_tests_dir', installed_tests_execdir)
     test_conf.set('program', test_name)
+    if installed_tests_env != {}
+      envs = []
+      foreach var, value : installed_tests_env
+         envs += '@0@=@1@'.format(var, value)
+       endforeach
+       test_conf.set('env', '@0@ @1@ '.format(env_program.path(), ' '.join(envs)))
+    endif
     configure_file(
       input: installed_tests_template_tap,
       output: test_name + '.test',
diff --git a/meson.build b/meson.build
index 98907ca..ad8c4cd 100644
--- a/meson.build
+++ b/meson.build
@@ -1893,6 +1893,9 @@ endif
 have_bash = find_program('bash', required : false).found() # For completion scripts
 have_sh = find_program('sh', required : false).found() # For glib-gettextize
 
+# Some installed tests require a custom environment
+env_program = find_program('env', required: installed_tests_enabled)
+
 # FIXME: How to detect Solaris? https://github.com/mesonbuild/meson/issues/1578
 if host_system == 'sunos'
   glib_conf.set('_XOPEN_SOURCE_EXTENDED', 1)
diff --git a/template-tap.test.in b/template-tap.test.in
index 6adf73f..30cd166 100644
--- a/template-tap.test.in
+++ b/template-tap.test.in
@@ -1,4 +1,4 @@
 [Test]
 Type=session
-Exec=@installed_tests_dir@/@program@ --tap
+Exec=@env@@installed_tests_dir@/@program@ --tap
 Output=TAP
